File: interviews/leetcode/stack/2390. Removing Stars From a String/solution.go
package main

func removeStars(s string) string {
	stack := make([]rune, 0, len(s))

	// Iterate through each character in the string
	for _, ch := range s {
		if ch == '*' {
			// If character is '*', pop the last character from the stack (the closest non-star character)
			stack = stack[:len(stack)-1]
		} else {
			// If character is not '*', push it onto the stack
			stack = append(stack, ch)
		}
	}

	// Convert the stack back to a string and return it
	return string(stack)
}

/*
func main() {
	// Example usage
	fmt.Println(removeStars("ab*c*d*ef")) // Output: "ef"
	fmt.Println(removeStars("****"))      // Output: ""
	fmt.Println(removeStars("abc*def**")) // Output: "ad"
}
*/

File: interviews/leetcode/stack/2390. Removing Stars From a String/solution.py
class Solution:
    def removeStars(self, s: str) -> str:
        stack = []

        # Iterate through each character in the string
        for ch in s:
            if ch == '*':
                # If character is '*', pop the last character from the stack
                stack.pop()
            else:
                # If character is not '*', add it to the stack
                stack.append(ch)

        # Convert the stack back to a string and return it
        return ''.join(stack)

# Example usage
# print(remove_stars("ab*c*d*ef"))  # Output: "ef"
# print(remove_stars("****"))       # Output: ""
# print(remove_stars("abc*def**"))  # Output: "ad"
